[UENV] CP2K: compilation and performance tests

RMeli opened the pull request:
CP2K compilation with the `develop` view, and performance tests for the executable provided by the UENV.

* [x] Parametrise compilation test on the version in the UENV
* [ ] ~Explore integration of full unit tests~
* [x] Compilation with `modules` view?

RMeli commented on checks/apps/cp2k/cp2k_check.py:
```diff
@@ -3,149 +3,227 @@
+    @run_before("run")
+    def prepare_run(self):
+        super().prepare_run()
+        self.job.options += [
+            "--time 00:05:00",
+        ]
```
This doesn't work...
```
[...]
#SBATCH --time 0:10:0
[...]
#SBATCH --time 00:15:00
[...]
```
The default time is still present.

RMeli commented on checks/apps/cp2k/cp2k_check.py:
```diff
@@ -3,149 +3,208 @@
+    def validate_test(self):
+        # INFO: Executables are in exe/local_cuda because -DCP2K_ENABLE_REGTEST=ON
+        # INFO: With -DCP2K_ENABLE_REGTEST=OFF, executables are in build/bin/ as expected
+        executable = os.path.join(self.stagedir, "exe", "local_cuda", "cp2k.psmp")
```
Folder `local_cuda` needs to be conditional on `uarch`.

RMeli commented on checks/apps/cp2k/cp2k_check.py:
```diff
@@ -3,149 +3,208 @@
-    @run_before('run')
-    def set_cpu_binding(self):
-        self.job.launcher.options = ['--cpu-bind=cores']
+class Cp2kCheckPBE(Cp2kCheck):
```
Run PBE test with built executable too? `+cp2k-dev` environment feature.

RMeli commented:
- [ ] Add `README.md` with file provenance
- [ ] Implement dependencies in a cleaner way
- [ ] Use shorter tests

jgphpc commented on checks/apps/cp2k/cp2k_check.py:
```diff
@@ -3,149 +3,208 @@
+import uenv
+
+cp2k_references = {
+    "pbe": {"gh200": {"time_run": (65, None, 5, "s")}},
```
Do you expect an upper limit of 65*5 or reference +5% ?
```suggestion
    "pbe": {"gh200": {"time_run": (65, None, 5/100, "s")}},
```

jgphpc commented on checks/apps/cp2k/src/BASIS_H2O:
```diff
@@ -0,0 +1,244 @@
+#
```
add a readme.md to show the origin of the input files

jgphpc commented on checks/apps/cp2k/cp2k_check.py:
```diff
@@ -3,149 +3,208 @@
 #
```
update the Copyright to 2016-2024

jgphpc commented on checks/apps/cp2k/cp2k_check.py:
```diff
@@ -3,149 +3,208 @@
+    valid_prog_environs = ["+cp2k-dev"]
+    build_system = "CMake"
+    sourcesdir = None
+    maintainers = ["RMeli"]
```
```suggestion
    maintainers = ["SSA"]
```

jgphpc commented on checks/apps/cp2k/cp2k_check.py:
```diff
@@ -3,149 +3,208 @@
-        else:
-            self.valid_prog_environs = ['builtin']
+    executable = "./mps-wrapper.sh cp2k.psmp"
+    maintainers = ["RMeli"]
```
```suggestion
    maintainers = ["SSA"]
```

jgphpc commented on checks/apps/cp2k/cp2k_check.py:
```diff
@@ -3,149 +3,208 @@
+    def prepare_build(self):
+        self.uarch = uenv.uarch(self.current_partition)
+        self.build_system.builddir = os.path.join(self.stagedir, "build")
+        self.build_system.max_concurrency = 64
```
```suggestion
        self.skip_if_no_procinfo()
        cpu = self.current_partition.processor
        self.build_system.max_concurrency = cpu.info['num_cpus_per_socket']
```

jgphpc commented:
CP2K is deployed on eiger:

```
uenv image find 
cp2k/2024.3:v1                          zen2  2024-09-18 7c7369b64b5fabe5 2.7GB     
cp2k/2024.1:v1                          zen2  2024-07-01 2a56f1df31a4c196 2.6GB  
```
hence we should either make the test work on eiger too or ignore eiger. What do you think ?

jgphpc commented on checks/apps/cp2k/cp2k_check.py:
```diff
@@ -3,149 +3,208 @@
-    modules = ['CP2K']
-    executable = 'cp2k.psmp'
-    executable_opts = ['H2O-256.inp']
-    maintainers = ['LM']
```
see comment below

jgphpc reviewed (changes requested):
see comments
(Tested with uenv/5.2.0-dev, UENV=cp2k/2024.3:v1 and 
[reframe.yaml](https://raw.githubusercontent.com/eth-cscs/alps-uenv/9a4367a350fb503b87c2afa7bdd7ec344f4a70cf/recipes/cp2k/2024.3/gh200/extra/reframe.yaml))

RMeli commented:
Thanks @jgphpc for all the comments. I'll definitely add Eiger, and tackle the remaining TODO items.

RMeli commented:
> we should either make the test work on eiger too or ignore eiger

It would be good to make the tests work on `eiger` too if the UENV is deployed there, but currently `eiger` has a different version of the SLURM plugin. Let's ignore it for now, and re-visit at a later stage.

teojgo reviewed (changes requested):
My only comment would be to change to single quote strings when possible to be consistent with the other tests.

RMeli commented:
Thanks, I'll do that. However, if we are interested in consistency down to quotes, I'd suggest to add a configuration file for some automated code formatter (eg `black`/`ruff`) so that everything is taken care of automagically.

jgphpc commented on checks/apps/cp2k/cp2k_check.py:
```diff
@@ -1,151 +1,230 @@
+        # SBATCH options
+        self.job.options = [
+            f"--nodes={config['nodes']}",
+            "--ntasks-per-core=1",
```
```suggestion
    self.num_tasks_per_core = 1
```

jgphpc commented on checks/apps/cp2k/cp2k_check.py:
```diff
@@ -1,151 +1,230 @@
+        self.job.options = [
+            f"--nodes={config['nodes']}",
+            "--ntasks-per-core=1",
+            "--partition=debug",
```
This is not portable so I would remove it and call reframe with `-J partition=debug` (or add it in the config file)

RMeli replied:
Yes, this is for me to test locally. I'll remove it before merging.

RMeli replied:
Done, I added the debug partition to the configuration and removed any reference to the system from the tests. I then used `--system todi:debug` run the tests.

RMeli commented:
I added tests for compilation with the `modules` view and updated the `reframe.yaml` in https://github.com/eth-cscs/alps-uenv/pull/156 accordingly: https://github.com/cp2k/cp2k/tree/support/v2024.3/benchmarks/QS_mp2_rpa/128-H2O

RMeli commented on config/systems/todi.py:
```diff
@@ -76,6 +76,46 @@
+            'scheduler': 'squeue',
+            'time_limit': '10m',
+            'container_platforms': [],
+            'access': ['--partition=debug'],
```
Needs to be removed?

teojgo replied:
Yes, I would say remove it for the moment and we can add it later in case we need.

RMeli commented on config/systems/todi.py:
```diff
@@ -76,6 +76,46 @@
+        {
+            'name': 'debug',
+            'descr': 'GH200',
+            'scheduler': 'squeue',
```
```suggestion
            'scheduler': 'slurm',
```
Using `squeue` circumvents `sacct` issues on `todi-ln001`.

RMeli commented:
I added an additional short MD test, and tested the whole pipeline with `'access': ['--reservation=dain']`. Everything completed successfully.

jgphpc reviewed (approved):
lgetm
